### PR TITLE
Add thrift protocol documents

### PR DESCRIPTION
**Pull request by kenshinxf**
Add thrift protocol documents

### CODE REVIEW

File: docs/en/thrift.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+[中文版](../cn/thrift.md)
+
+[thrift](https://thrift.apache.org/)is a RPC framework used widely in production area which was developed by Facebook, In order to access thrift servers more conveniently and make full use of bthread's capability of concurrency, brpc directly supports the thrift protocol.
```
Comment (zyearn) [inline docs/en/thrift.md]: production area ->  production environment
Facebook, -> Facebook.
bthread's capability of concurrency -> concurrency ability of bthread

File: docs/en/thrift.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
+[中文版](../cn/thrift.md)
+
+[thrift](https://thrift.apache.org/)is a RPC framework used widely in production area which was developed by Facebook, In order to access thrift servers more conveniently and make full use of bthread's capability of concurrency, brpc directly supports the thrift protocol.
+Check [example/thrift_extension_c++](https://github.com/brpc/brpc/tree/master/example/thrift_extension_c++/) for an example.
+
+Advantages compared to the official thrift client:
+- Thread safety. No need to set up separate clients for each thread.
+- Support synchronous, asynchronous, semi-synchronous accesses etc. Support [ParallelChannel etc](combo_channel.md) to define access patterns declaratively.
```
Comment (zyearn) [inline docs/en/thrift.md]: 和中文对不上?
Supports synchronous, asynchronous, batch synchronous, batch asynchronous, and other access methods. Combination channels such as ParallelChannel are also supported.

File: docs/en/thrift.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+[thrift](https://thrift.apache.org/)is a RPC framework used widely in production area which was developed by Facebook, In order to access thrift servers more conveniently and make full use of bthread's capability of concurrency, brpc directly supports the thrift protocol.
+Check [example/thrift_extension_c++](https://github.com/brpc/brpc/tree/master/example/thrift_extension_c++/) for an example.
+
+Advantages compared to the official thrift client:
+- Thread safety. No need to set up separate clients for each thread.
+- Support synchronous, asynchronous, semi-synchronous accesses etc. Support [ParallelChannel etc](combo_channel.md) to define access patterns declaratively.
+- Support various connection types(short, connection pool). Support timeout, backup request, cancellation, tracing, built-in services, and other benefits offered by brpc.
+
+# Compile and Run
+In order to not depend on and compile with thrift library for most of the users, the thrift protocol wasn't supported in brpc by default. configure brpc with --with-thrift if you want to enable thrift protocol in brpc(actually it works with a Macro ENABLE_THRIFT_FRAMED_PROTOCOL)
```
Comment (zyearn) [inline docs/en/thrift.md]: for most of the users -> for most users
configure -> Configure
works with -> works by using

File: docs/en/thrift.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+cd thrift-0.9.3/
+./configure --prefix=/usr --with-ruby=no --with-python=no --with-java=no --with-go=no --with-perl=no --with-php=no --with-csharp=no --with-erlang=no --with-lua=no --with-nodejs=no
+make -j 3 -s
+sudo make install
+```
+Configure with thrift support
+```bash
+sh config_brpc.sh --headers=/usr/include --libs=/usr/lib --nodebugsymbols --with-thrift
+```
+Link with libbrpc_thrift.a if the user want to enable thrift protocol
```
Comment (zyearn) [inline docs/en/thrift.md]: the user -> users
protocol -> protocol.
另外似乎和中文对不上

File: docs/en/thrift.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+brpc::ChannelOptions options;
+options.protocol = brpc::PROTOCOL_THRIFT;
+brpc::Channel thrift_channel;
+if (thrift_channel.Init(Flags_server.c_str(), FLAGS_load_balancer.c_str(), &options) != 0) {
+   LOG(ERROR) << "Fail to initialize thrift channel";
+   return -1;
+}
+...
+```
+construct thrift request and send to server, ThriftTemplateMessage is a template class, the native thrift message was delegated by and can be accessed directly by raw() method.
```
Comment (zyearn) [inline docs/en/thrift.md]: Construct a thrift request and send it to server. ThriftTemplateMessage is a template class that hosts thrift native messages. raw() methods can directly manipulate native thrift messages.

File: docs/en/thrift.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+ channel.CallMethod(NULL, &cntl, &req, &res, NULL);
+
+ if (cntl.Failed()) {
+     LOG(ERROR) << "Fail to send thrift request, " << cntl.ErrorText();
+     return -1;
+ } 
+```
+
+# Server process with thrift message
```
Comment (zyearn) [inline docs/en/thrift.md]: Server side processing upstream thrfit requests

File: docs/en/thrift.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+User need to implement it's own thrift handler which was inherited from brpc::ThriftService
```
Comment (zyearn) [inline docs/en/thrift.md]: User ->Users
it's -> their
brpc::ThriftService -> brpc::ThriftService.

File: docs/en/thrift.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+Only the method name can obtained in server side due to the limit of thrift protocol itself, the service name wasn't passed to server actually, it means that only one thrift service can be set in one brpc server
```
Comment (zyearn) [inline docs/en/thrift.md]: Due to the limitation of thrift protocol itself, only the method name can be obtained on the server (via the controller.thrift_method_name() method), and the service name cannot be obtained. This is consistent with the native thrift implementation, which means that there can only be one thrift service in one brpc server.

File: docs/en/thrift.md (hunk 1 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: Thanks, Updated.

File: docs/en/thrift.md (hunk 2 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: done

File: docs/en/thrift.md (hunk 3 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: done

File: docs/en/thrift.md (hunk 4 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: done

File: docs/en/thrift.md (hunk 7 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: done

File: docs/en/thrift.md (hunk 6 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: done

File: README.md [hunk 9]
```diff
@@ -13,7 +13,7 @@ You can use it to:
   * [rtmp](https://github.com/brpc/brpc/blob/master/src/brpc/rtmp.h)/[flv](https://en.wikipedia.org/wiki/Flash_Video)/[hls](https://en.wikipedia.org/wiki/HTTP_Live_Streaming), for building [live-streaming services](docs/cn/live_streaming.md).
   * hadoop_rpc (may be opensourced)
   * [rdma](https://en.wikipedia.org/wiki/Remote_direct_memory_access) support (will be opensourced)
-  * all sorts of protocols used in Baidu: [baidu_std](docs/cn/baidu_std.md), [streaming_rpc](docs/en/streaming_rpc.md), hulu_pbrpc, [sofa_pbrpc](https://github.com/baidu/sofa-pbrpc), nova_pbrpc, public_pbrpc, ubrpc, and nshead-based ones.
+  * all sorts of protocols used in Baidu: [baidu_std](docs/cn/baidu_std.md), [streaming_rpc](docs/en/streaming_rpc.md), hulu_pbrpc, [sofa_pbrpc](https://github.com/baidu/sofa-pbrpc), nova_pbrpc, public_pbrpc, ubrpc, [thrift](docs/en/thrift.md) and nshead-based ones.
```
[the remark this answers is not in the document]
Comment (kenshinxf) [reply README.md]: done

File: docs/en/thrift.md (hunk 8 above)
Comment (kenshinxf) [reply docs/en/thrift.md]: Thanks for all of the comments above, :)

File: README_cn.md [hunk 10]
```diff
@@ -14,7 +14,7 @@
   * [rtmp](https://github.com/brpc/brpc/blob/master/src/brpc/rtmp.h)/[flv](https://en.wikipedia.org/wiki/Flash_Video)/[hls](https://en.wikipedia.org/wiki/HTTP_Live_Streaming), 可用于搭建[直播服务](docs/cn/live_streaming.md).
   * hadoop_rpc(可能开源)
   * 支持[rdma](https://en.wikipedia.org/wiki/Remote_direct_memory_access)(即将开源)
-  * 各种百度内使用的协议: [baidu_std](docs/cn/baidu_std.md), [streaming_rpc](docs/cn/streaming_rpc.md), hulu_pbrpc, [sofa_pbrpc](https://github.com/baidu/sofa-pbrpc), nova_pbrpc, public_pbrpc, ubrpc和使用nshead的各种协议.
+  * 各种百度内使用的协议: [baidu_std](docs/cn/baidu_std.md), [streaming_rpc](docs/cn/streaming_rpc.md), hulu_pbrpc, [sofa_pbrpc](https://github.com/baidu/sofa-pbrpc), nova_pbrpc, public_pbrpc, ubrpc, [thrift](docs/cn/thrift.md)和使用nshead的各种协议.
```
[the remark this answers is not in the document]
Comment (kenshinxf) [reply README_cn.md]: done

File: docs/cn/thrift.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
+[English Version](../en/thrift.md)
+
+[thrift](https://thrift.apache.org/)是近几年应用较广的Facebook发布的RPC服务, 为了使用户更方便,快捷的利用bthread的并发能力，brpc实现并支持thrift工作在NonBlocking模式下的协议(FramedProtocol), 注意本文中所说的thrift协议一律指的是此种情况下的thrift协议.
```
[the remark this answers is not in the document]
Comment (kenshinxf) [reply docs/cn/thrift.md]: Thrift 官方文档不是特别好, 关于协议这块暂时还没找到合适的链接

File: docs/cn/thrift.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+示例程序：[example/thrift_extension_c++](https://github.com/brpc/brpc/tree/master/example/thrift_extension_c++/)
+
+相比使用官方原生的优势有：
+
+- 线程安全。用户不需要为每个线程建立独立的client。
+- 支持同步、异步、批量同步、批量异步等访问方式，能使用ParallelChannel等组合访问方式。
+- 支持多种连接方式(连接池, 短连接), 支持超时、backup request、取消、tracing、内置服务等一系列RPC基本福利。
+
+# 编译依赖及运行
+默认brpc编译是不启用thrift协议支持的, 目的是在用户不需要thrift协议支持的情况下可以不安装thrift依赖. 如果用户启用thrift协议的话, 在配置brpc环境的时候加上--with-thrift参数(实际上是在Makefile里面启用ENABLE_THRIFT_FRAMED_PROTOCOL宏)
```
[the remark this answers is not in the document]
Comment (kenshinxf) [reply docs/cn/thrift.md]: done

File: docs/cn/thrift.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+安装thrift依赖, ubuntu环境下
```
[the remark this answers is not in the document]
Comment (kenshinxf) [reply docs/cn/thrift.md]: done

**Pull request merged**